### PR TITLE
Adding anti lag

### PR DESCRIPTION
**Pull request by JacobHolman**
prevents:
over 100 mobs at once (value in config)
max water flow maxed at 150 (in config)
torch limit cap is 1000 (config)
stops common crash methods of flooding the server
which crashed the server multiple times

### CODE REVIEW

**Comment by LunaWasFlaggedAgain** [comment]
There already is an entity limiter built into Extras. Consider making a PR to lower the entity limit amount from 1500 to another value if you're not happy with the current limit.

The torch limiter can easily be bypassed with commands such as /fill and using WorldEdit. It also doesn't work properly, as you're limiting torch placement based on the count of torches in someone inventory, and not the amount of torches they've placed.

The water flow count is never reset, so after enough water flow events the server will stop processing them until it's restarted. The same thing goes for the entity limiter.

You never updated the default config to have the new values you've added, so every limit will be set to zero by default, making the server broken, as people won't be able to spawn entities, get water to flow or place torches.

**Comment by RealLava** [comment]
i forgot to commit the config update

**Review by allinkdev — CHANGES_REQUESTED** [review]
Please add the configuration values you use in this PR to config.yml.

File: src/main/java/pw/kaboom/extras/modules/checks/DenyMobLag.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import pw.kaboom.extras.Main;
+
+public class DenyMobLag implements Listener {
+    private static final int limit = JavaPlugin.getPlugin(Main.class).getConfig().getInt("mobLimitAtOnce");
+    private int mobSpawnCount = 0;
+
+    @EventHandler
+    public void onCreatureSpawn(CreatureSpawnEvent event) {
+        EntityType entityType = event.getEntityType();
+        if (entityType.isAlive()) {
```
Comment (allinkdev) [inline src/main/java/pw/kaboom/extras/modules/checks/DenyMobLag.java:17]: This .isAlive() check is unnecessary in a CreatureSpawnEvent listener, given that the CreatureSpawnEvent will only fire for LivingEntities.

File: src/main/java/pw/kaboom/extras/modules/checks/DenyMobLag.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import org.bukkit.entity.EntityType;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.plugin.java.JavaPlugin;
+import pw.kaboom.extras.Main;
+
+public class DenyMobLag implements Listener {
+    private static final int limit = JavaPlugin.getPlugin(Main.class).getConfig().getInt("mobLimitAtOnce");
+    private int mobSpawnCount = 0;
```
Comment (allinkdev) [inline src/main/java/pw/kaboom/extras/modules/checks/DenyMobLag.java:12]: This counter will never deplete, meaning that when the limit is exceeded, no mobs will be able to spawn, even if the server is well below the mob capacity limit.

File: src/main/java/pw/kaboom/extras/modules/checks/DenyTorchLoad.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import org.bukkit.entity.Player;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.plugin.java.JavaPlugin;
+import pw.kaboom.extras.Main;
+
+public class DenyTorchLoad implements Listener {
+    private static final int limit = JavaPlugin.getPlugin(Main.class).getConfig().getInt("torchLimit");
+
+    @EventHandler
+    public void onBlockPlace(BlockPlaceEvent event) {
```
Comment (allinkdev) [inline src/main/java/pw/kaboom/extras/modules/checks/DenyTorchLoad.java:16]: This entire event listener will never work. The amount of torches in a player's inventory is in no way proportional to how many torches are in the world/chunk. This event will do nothing but confuse players.

File: src/main/java/pw/kaboom/extras/modules/checks/DenyWaterFlow.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import org.bukkit.Material;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.block.BlockFromToEvent;
+import org.bukkit.plugin.java.JavaPlugin;
+import pw.kaboom.extras.Main;
+
+public class DenyWaterFlow implements Listener {
+    private static final int max = JavaPlugin.getPlugin(Main.class).getConfig().getInt("maxWaterFlow");
+    private int waterFlowCount = 0;
```
Comment (allinkdev) [inline src/main/java/pw/kaboom/extras/modules/checks/DenyWaterFlow.java:12]: The liquid flow counter is never depleted, meaning that if the server hits the limit, water and lava will never be able to flow again, even if the server is well within the limits.

File: src/main/java/pw/kaboom/extras/modules/checks/DenyTorchLoad.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.block.BlockPlaceEvent;
+import org.bukkit.entity.Player;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.plugin.java.JavaPlugin;
+import pw.kaboom.extras.Main;
+
+public class DenyTorchLoad implements Listener {
+    private static final int limit = JavaPlugin.getPlugin(Main.class).getConfig().getInt("torchLimit");
```
Comment (allinkdev) [inline src/main/java/pw/kaboom/extras/modules/checks/DenyTorchLoad.java:13]: You are getting configuration values statically. This is not ideal, as it makes it impossible to change configuration values without restarting the server.

File: src/main/java/pw/kaboom/extras/modules/checks/DenyWaterFlow.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+import org.bukkit.Material;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.block.BlockFromToEvent;
+import org.bukkit.plugin.java.JavaPlugin;
+import pw.kaboom.extras.Main;
+
+public class DenyWaterFlow implements Listener {
+    private static final int max = JavaPlugin.getPlugin(Main.class).getConfig().getInt("maxWaterFlow");
```
Comment (allinkdev) [inline src/main/java/pw/kaboom/extras/modules/checks/DenyWaterFlow.java:11]: You are also getting configuration values statically here.

File: src/main/java/pw/kaboom/extras/modules/checks/DenyWaterFlow.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    private int waterFlowCount = 0;
+
+    @EventHandler
+    public void onBlockFromTo(BlockFromToEvent event) {
+        if (!JavaPlugin.getPlugin(Main.class).getConfig().getBoolean("blockLotsOfWaterFromFlowing")) return;
```
Comment (allinkdev) [inline src/main/java/pw/kaboom/extras/modules/checks/DenyWaterFlow.java:16]: Why is this configuration value the exception to the static rule? All others in your pull request are done in the static constructor.

File: src/main/java/pw/kaboom/extras/modules/checks/DenyMobLag.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
+package pw.kaboom.extras.modules.checks;
+
+import org.bukkit.entity.EntityType;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.plugin.java.JavaPlugin;
+import pw.kaboom.extras.Main;
+
+public class DenyMobLag implements Listener {
```
Comment (allinkdev) [inline src/main/java/pw/kaboom/extras/modules/checks/DenyMobLag.java:10]: Extras already has a mob limiter. Consider removing this class or merging it into the current mob limiter implementation.

**Review by allinkdev — CHANGES_REQUESTED** [review]

**Pull request closed without merging**